admin/projects.go

```diff
@@ -118,6 +119,208 @@ func (s *Service) TeardownProject(ctx context.Context, p *database.Project) erro
+			Description:            proj.Description,
+			Public:                 proj.Public,
+			ProductionBranch:       proj.ProductionBranch,
+			ProductionVariables:    proj.ProductionVariables,
+			GithubURL:              proj.GithubURL,
+			GithubInstallationID:   proj.GithubInstallationID,
+			ProductionDeploymentID: &depl.ID,
+		})
+		if err != nil {
+			err2 := s.DB.DeleteDeployment(ctx, depl.ID)
+			err3 := s.provisioner.Teardown(ctx, inst.Host, inst.InstanceID, proj.ProductionOLAPDriver)
+			err4 := s.DB.DeleteProject(ctx, proj.ID)
+			s.logger.Error("trigger redeploy: could not update project", zap.String("instance_id", inst.InstanceID), zap.Error(multierr.Combine(err, err2, err3, err4)))
+			return
+		}
+
+		// Trigger reconcile
+		err = s.TriggerReconcile(ctx, depl.ID)
+		if err != nil {
+			// This error is weird. But it's safe not to teardown the rest.
+			s.logger.Error("trigger redeploy: could not reconcile project", zap.String("deployment_id", depl.ID), zap.Error(err))
+			return
+		}
+	}()
+	return nil
+}
+
+// New method for refresh source
+func (s *Service) TriggerRefreshSource(ctx context.Context, deploymentID string) error {
+	// Run it all in the background
+	go func() {
+		// Use s.closeCtx to cancel if the service is stopped
+		ctx := s.closeCtx
+
+		s.logger.Info("refresh source: starting", zap.String("deployment_id", deploymentID))
+
+		// Get deployment
+		depl, err := s.DB.FindDeployment(ctx, deploymentID)
+		if err != nil {
+			s.logger.Error("refresh source: could not find deployment", zap.String("deployment_id", deploymentID), zap.Error(err))
+			return
+		}
+
+		// Check status
+		if depl.Status == database.DeploymentStatusReconciling && time.Since(depl.UpdatedOn) < 30*time.Minute {
+			s.logger.Error("refresh source: skipping because it is already running", zap.String("deployment_id", deploymentID))
+			return
+		}
+
+		// Set deployment status to reconciling
+		depl, err = s.DB.UpdateDeploymentStatus(ctx, deploymentID, database.DeploymentStatusReconciling, "")
+		if err != nil {
+			s.logger.Error("refresh source: could not update status", zap.String("deployment_id", deploymentID), zap.Error(err))
+			return
+		}
+
+		// Get superuser token for runtime host
+		jwt, err := s.issuer.NewToken(auth.TokenOptions{
+			AudienceURL:         depl.RuntimeAudience,
+			TTL:                 time.Hour,
+			InstancePermissions: map[string][]auth.Permission{depl.RuntimeInstanceID: {auth.EditInstance}},
+		})
+		if err != nil {
+			s.logger.Error("refresh source: could not get token", zap.String("deployment_id", deploymentID), zap.Error(err))
+			return
+		}
+
+		// Make runtime client
+		rt, err := client.New(depl.RuntimeHost, jwt)
+		if err != nil {
+			s.logger.Error("refresh source: could not create client", zap.String("deployment_id", deploymentID), zap.Error(err))
+			return
+		}
+
+		// Call refresh source
+		res, err := rt.TriggerRefresh(ctx, &runtimev1.TriggerRefreshRequest{InstanceId: depl.RuntimeInstanceID})
```

Comment:
It's not passing a name? This is an invalid request

Reply:
oh yes, missed it.